File: leetcode/leetcode/binary_tree_level_order_traversal.go
package leetcode

import "fmt"

func LevelOrder() {
	{
		three := &TreeNode[int]{Val: 3}
		nine := &TreeNode[int]{Val: 9}
		twenty := &TreeNode[int]{Val: 20}
		fifteen := &TreeNode[int]{Val: 15}
		seven := &TreeNode[int]{Val: 7}

		three.Left = nine
		three.Right = twenty
		twenty.Left = fifteen
		twenty.Right = seven

		fmt.Println(levelOrder(three))
	}

	{
		one := &TreeNode[int]{Val: 1}

		fmt.Println(levelOrder(one))
	}

	{
		fmt.Println(levelOrder(nil))
	}
}

func levelOrder(root *TreeNode[int]) [][]int {
	if root == nil {
		return [][]int{}
	}

	ans := [][]int{}
	q := &Queue[*TreeNode[int]]{}
	q.Enqueue(root)

	for q.Len() > 0 {
		n := q.Len()
		curr_ans := []int{}
		for i := 0; i < n; i++ {
			node := q.Dequeue()
			curr_ans = append(curr_ans, node.Val)
			if node.Left != nil {
				q.Enqueue(node.Left)
			}
			if node.Right != nil {
				q.Enqueue(node.Right)
			}
		}
		ans = append(ans, curr_ans)
	}

	return ans
}

File: leetcode/leetcode/binary_tree_level_order_traversal.js
/**
 * @param {number} val
 * @param {TreeNode} left
 * @param {TreeNode} right
 */
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  if (root == null) {
    return [];
  }

  /** @type {[TreeNode, level][]} */
  const q = [];
  let curr = root;
  q.push([curr, 0]);

  /** @type {number[][]} */
  const ans = [];

  while (q.length) {
    const nl = q.shift();
    if (nl && nl[0]) {
      const node = nl[0];
      const level = nl[1];
      ans[level] = ans[level] || [];
      ans[level].push(node.val);
      q.push([node.left, level + 1]);
      q.push([node.right, level + 1]);
    }
  }

  return ans;
};

{
  const three = new TreeNode(3);
  const nine = new TreeNode(9);
  const twenty = new TreeNode(20);
  const fifteen = new TreeNode(15);
  const seven = new TreeNode(7);

  three.left = nine;
  three.right = twenty;
  twenty.left = fifteen;
  twenty.right = seven;

  console.log(levelOrder(three));
}
